### PR TITLE
[ML] Add stream flag to inference providers

### PR DESCRIPTION
**Pull request by prwhelan**
Pass the stream flag from the REST request through to the inference providers via the InferenceInputs.

### CODE REVIEW

**Comment by prwhelan** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/inference/InferenceService.java [hunk 1]
```diff
@@ -94,6 +95,7 @@ void infer(
         Model model,
         @Nullable String query,
         List<String> input,
+        boolean stream,
```
Comment (dan-rubinstein) [inline server/src/main/java/org/elasticsearch/inference/InferenceService.java:98]: Is the stream boolean passed in with every request to perform an inference? Is this something that can be part of task settings instead?

File: x-pack/plugin/inference/qa/test-service-plugin/src/main/java/org/elasticsearch/xpack/inference/mock/TestDenseInferenceServiceExtension.java [hunk 2]
```diff
@@ -94,6 +94,7 @@ public void infer(
             Model model,
             @Nullable String query,
             List<String> input,
+            boolean stream,
```
Comment (dan-rubinstein) [inline x-pack/plugin/inference/qa/test-service-plugin/src/main/java/org/elasticsearch/xpack/inference/mock/TestDenseInferenceServiceExtension.java:97]: It seems that we have multiple inference service extensions which do not use streaming at all. Why is this?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/HttpUtils.java [hunk 3]
```diff
@@ -46,7 +46,7 @@ private static String getStatusCodeErrorMessage(Request request, HttpResult resu
     }
 
     public static void checkForEmptyBody(ThrottlerManager throttlerManager, Logger logger, Request request, HttpResult result) {
-        if (result.isBodyEmpty()) {
+        if (result.isBodyEmpty() && (request.isStreaming() == false)) {
```
Comment (dan-rubinstein) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/HttpUtils.java:49]: Why can a streaming request have an empty body?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/sender/QueryAndDocsInputs.java [hunk 4]
```diff
@@ -21,6 +21,19 @@ public static QueryAndDocsInputs of(InferenceInputs inferenceInputs) {
...
 
     private final String query;
+    private final List<String> chunks;
+    private final boolean stream;
+
+    public QueryAndDocsInputs(String query, List<String> chunks) {
+        this(query, chunks, false);
+    }
+
+    public QueryAndDocsInputs(String query, List<String> chunks, boolean stream) {
```
Comment (dan-rubinstein) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/sender/QueryAndDocsInputs.java:31]: Why are we calling the values chunks here and inputs in the DocumentsOnlyInputs?

File: server/src/main/java/org/elasticsearch/inference/InferenceService.java (hunk 1 above)
Comment (dan-rubinstein) [reply server/src/main/java/org/elasticsearch/inference/InferenceService.java:98]: Makes sense, I agree that this seems like the best option in the current system. Thanks for clarifying

File: x-pack/plugin/inference/qa/test-service-plugin/src/main/java/org/elasticsearch/xpack/inference/mock/TestDenseInferenceServiceExtension.java (hunk 2 above)
Comment (dan-rubinstein) [reply x-pack/plugin/inference/qa/test-service-plugin/src/main/java/org/elasticsearch/xpack/inference/mock/TestDenseInferenceServiceExtension.java:97]: Got it, thanks for clarifying.

**Review by dan-rubinstein — APPROVED** [review]

**Comment by prwhelan** [comment]
@elasticmachine update branch

**Pull request merged**